### PR TITLE
Check for AES-NI on non-Intel processors

### PR DESCRIPTION
**Pull request by Teknoman117**
### Description

Remove the check for an Intel CPU in intel_aes_instructions_present().  Only check the aes-ni bit
in cpuid.

### Motivation and Context

AMD CPUs since the FX series (Family 15h) have support for AES-NI.  This can be checked via
cpuid exactly the same way as Intel CPUs.  Addresses #7102 

### How Has This Been Tested?

tested on my personal desktop with a Threadripper 1950x.
ZFS version: 0.7.0-263_g0735ecb3
SPL version: 0.7.0-24_g23602fd

- have a zpool called test on a ramdisk (to try and make this compute-bound)
- zpool import -R /mnt test
- zfs create -o encryption=on -o keyformat=passphrase -o keylocation=prompt test/encrypted
- dd if=/dev/urandom of=/mnt/test/encrypted/file.bin bs=1024k count=4096
- zpool export test
- zpool import -R /mnt test
- zfs load-key test/encrypted
- zfs mount test/encrypted
- dd if=/mnt/test/encrypted/file.bin of=/dev/null status=progress

Performance reading file.bin
- ccm before change: 1318.02 MiB/s
- ccm after change: 2730.5 MiB/s
- gcm before change: 350 MiB/s
- gcm after change: 2900 MiB/s
- will run fio tomorrow

I still need to run the zfs tests, but i first need to create a vm of something to run them on or figure out how to run them on gentoo.
### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: module/icp/algs/aes/aes_impl.c [hunk 1]
```diff
@@ -1584,34 +1584,21 @@ intel_aes_instructions_present(void)
...
-			__asm__ __volatile__(
-			    "cpuid"
-			    : "=a" (eax), "=b" (ebx), "=c" (ecx), "=d" (edx)
-			    : "a"(func), "c"(subfunc));
-
-			cached_result = !!(ecx & INTEL_AESNI_FLAG);
-		} else {
-			cached_result = 0;
-		}
+	if (cached_result != -1) {
```
Comment (tcaputi) [inline module/icp/algs/aes/aes_impl.c]: (Nitpick) Remove the brackets

Comment (kpande) [reply module/icp/algs/aes/aes_impl.c]: don't check for Intel CPU. this would affect virtualised instances.

**Comment by kpande** [comment]
please don't explicitly check for amd or Intel CPU. this would affect virtualised instances.

**Comment by tcaputi** [comment]
> please don't explicitly check for amd or Intel CPU. this would affect virtualised instances.

My only concern here is if different manufacturers use bit 25 to mean something other than AES-NI. That would almost definitely lead to a kernel panic.

**Comment by rlaager** [comment]
@kpande From what I can see, the commit removes the explicit check for Intel and it does not add an explicit check for AMD.

**Comment by kpande** [comment]
@rlaager see @tcaputi 's comments above asking for clarification on removing entirely vs adding one for AMD.

File: module/icp/algs/aes/aes_impl.c [hunk 2]
```diff
@@ -1570,7 +1570,7 @@ aes_alloc_keysched(size_t *size, int kmflag)
 
 #ifdef __amd64
 
-#define	INTEL_AESNI_FLAG (1 << 25)
+#define	CPUID_AESNI_FLAG (1 << 25)
```
Comment (behlendorf) [inline module/icp/algs/aes/aes_impl.c]: The comment above `intel_aes_instructions_present` needs to be updated.

**Comment by rlaager** [comment]
@kpande Sorry, I missed that.

It's unclear to me how CPUID flags work. Wikipedia links to this version of Linux's cpufeature.h, which describes it in the Intel block: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/x86/include/asm/cpufeature.h?id=61c4f2c81c61f73549928dfd9f3e8f26aa36a8cf

That code has been re-worked since that version. It seems that the current approach is to call boot_cpu_has(X86_FEATURE_AES). That seems to be GPL.

File: module/icp/algs/aes/aes_impl.c [hunk 3]
```diff
@@ -1584,34 +1584,21 @@ intel_aes_instructions_present(void)
...
-			func = 1;
-			subfunc = 0;
-
-			/* check for aes-ni instruction set */
-			__asm__ __volatile__(
-			    "cpuid"
-			    : "=a" (eax), "=b" (ebx), "=c" (ecx), "=d" (edx)
-			    : "a"(func), "c"(subfunc));
-
-			cached_result = !!(ecx & INTEL_AESNI_FLAG);
```
Comment (behlendorf) [inline module/icp/algs/aes/aes_impl.c]: I'd recommend that we update this function to call the CPU feature checking code which was added for vectorized RAIDZ in `include/linux/simd_x86.h`.  It will need to be extended slightly to check for the ` aes-ni ` instruction but then all of this code will be in a single place.

**Comment by Teknoman117** [comment]
@behlendorf I'll get on it when I get back from work today.

**Comment by behlendorf** [comment]
> That code has been re-worked since that version. It seems that the current approach is to call boot_cpu_has(X86_FEATURE_AES). That seems to be GPL.

Yes, this is exactly what the RAIDZ code uses, although I don't recall any GPL-only issues.

**Comment by ironMann** [comment]
As suggested, please try to extend the existing code in `config/toolchain-simd.m4` and `include/linux/simd_x86.h`. This will automatically use the kernel's GPL interface when available, and fallback to our own feature detection when not. (see `zfs_avx_available()` how an example)
The test will also work in userspace if needed (e.g. ztest).

**Comment by tcaputi** [comment]
@Teknoman117 
Thanks for submitting this PR. Would you mind making one small addition while we're doing this? `intel_pclmulqdq_instruction_present()` in `module/icp/algs/modes/gcm.c` should probably also be adjusted to use the logic discussed above.

**Comment by Teknoman117** [comment]
@tcaputi @behlendorf @ironMann I've added the toolchain checks for the instruction sets and populated simd_x86 with the feature flag checks as requested.  I do have a question regarding `module/icp/algs/aes/aes_impl.c` though.  It currently has a flags field where it stores the result of the feature flag check that gets passed to the crypto functions.  Would it be preferable to change it to the function pointer pattern used by much of the raidz code?

File: module/icp/algs/aes/aes_impl.c [hunk 4]
```diff
@@ -951,15 +954,18 @@ rijndael_key_setup_enc_raw(uint32_t rk[], const uint32_t cipherKey[],
 static void
 aes_setupkeys(aes_key_t *key, const uint32_t *keyarr32, int keybits)
 {
-	if (intel_aes_instructions_present()) {
+#if defined(HAVE_AES)
+	if (zfs_aes_available()) {
 		key->flags = INTEL_AES_NI_CAPABLE;
```
Comment (Teknoman117) [inline module/icp/algs/aes/aes_impl.c]: Should this be changed to use the function pointer model that the vdev raidz code uses?

Comment (rlaager) [reply module/icp/algs/aes/aes_impl.c]: What's the test situation like here? If the implementations are selectable, then it's possible to run the tests against the various implementations (e.g. with and without AES-NI).

File: module/icp/algs/aes/aes_impl.c (hunk 3 above)
Comment (Teknoman117) [reply module/icp/algs/aes/aes_impl.c]: I've now addressed this

**Comment by Teknoman117** [comment]
Figured I'd mention that using PCLMULQDQ on Ryzen increased the sequential read on aes-128-gcm from ~350 MiB/s to 2900 MiB/s.

File: include/linux/simd_x86.h [hunk 5]
```diff
@@ -194,7 +196,9 @@ static const cpuid_feature_desc_t cpuid_features[] = {
 	[AVX512VBMI]	= {7U, 0U, _AVX512VBMI_BIT,	ECX	},
 	[AVX512PF]	= {7U, 0U, _AVX512PF_BIT,	EBX	},
 	[AVX512ER]	= {7U, 0U, _AVX512ER_BIT,	EBX	},
-	[AVX512VL]	= {7U, 0U, _AVX512ER_BIT,	EBX	}
+	[AVX512VL]	= {7U, 0U, _AVX512ER_BIT,	EBX	},
+	[AES]		= {1U, 0U,	1U << 25,	ECX	},
+	[PCLMULQDQ]	= {1U, 0U,	1U << 1,	ECX	},
```
Comment (behlendorf) [inline include/linux/simd_x86.h]: Please define `_AES_BIT` and `_PCLMULQDQ_BIT` above and use them here.

File: module/icp/algs/aes/aes_impl.c [hunk 6]
```diff
@@ -22,12 +22,13 @@
...
 #include <linux/simd_x86.h>
+#endif
 
+#if defined(_KERNEL) && defined(__amd64)
 #define	KPREEMPT_DISABLE	kfpu_begin()
 #define	KPREEMPT_ENABLE		kfpu_end()
-
 #else
 #define	KPREEMPT_DISABLE
 #define	KPREEMPT_ENABLE
```
Comment (behlendorf) [inline module/icp/algs/aes/aes_impl.c]: The `linux/simd_x86.h` header already takes care of defining `kfpu_begin/end` properly for the build environment.  You should be able to simply unconditionally define `KPREEMPT_DISABLE` to be `kfpu_begin()`.  I wouldn't be at all against a search and replace of `KPREEMPT_DISABLE` with `kfpu_begin()`, etc.  Even if it's a slight divergence from Illumos.

File: module/icp/algs/aes/aes_impl.c (hunk 4 above)
Comment (behlendorf) [reply module/icp/algs/aes/aes_impl.c]: That'd be a nice addition to facilitate testing and to add visibility to make it easy to verify you are using the accelerated version.  It also nicely resolves the issue of caching the result, so I think it would be a nice addition.

File: module/icp/algs/modes/gcm.c [hunk 7]
```diff
@@ -22,12 +22,13 @@
...
  */
 
-#if defined(_KERNEL) && defined(__amd64)
+#if defined(__amd64)
 #include <linux/simd_x86.h>
+#endif
 
+#if defined(_KERNEL) && defined(__amd64)
 #define	KPREEMPT_DISABLE	kfpu_begin()
 #define	KPREEMPT_ENABLE		kfpu_end()
```
Comment (behlendorf) [inline module/icp/algs/modes/gcm.c]: Same as above.

**Comment by behlendorf** [comment]
@Teknoman117 please let us know if you can't spare the time to finalize this.  I'm sure we can help wrap it up if you like.

**Comment by Teknoman117** [comment]
@behlendorf Had some stuff come up at work that took a lot of attention, I'll be able to continue on Monday.

**Comment by kpande** [comment]
@Teknoman117 this patch is rather urgent, any time to work on it?

**Comment by Teknoman117** [comment]
@kpande nearly done, will update monday evening.

**Comment by Teknoman117** [comment]
@behlendorf  @kpande i've made the requested changes with one exception.  At the moment, the aes implementation is checked per call.  I wanted to ask whether you wanted me to pass an ops struct from the aes mode implementation (ccm, gcm, ctr, etc.) as a parameter to aes_encrypt_block or to have a global pointer to it in aes_impl.

**Comment by behlendorf** [comment]
@Teknoman117 nice work, thanks again for tackling this.

Actually, I'd suggest we stick with calling `aes_impl_get_ops()`  for each call.  When doing the RAIDZ vectorization work we found that this extra call didn't have any measurable imact on overall performance.  I'd expect the same to be the case here, but I'd love to confirm that with real performance data.

Additionally, this allowed us to easily cycle through all the available methods on a per-block basis as part of the automated test coverage to rigerously verify compatibility.  This is the ` IMPL_CYCLE` case in the `vdev_raidz_math_get_ops()` function which is only defined in user space and is leverage by `ztest`.  We should do the same thing here, you're updated patch almost has everything needed for this in place.

**Comment by Teknoman117** [comment]
@behlendorf do you recall why the cycle logic is guarded by #if !defined(_KERNEL) ?

**Comment by behlendorf** [comment]
@Teknoman117 there's no functional reason this couldn't be allowed this in the kernel.  However, the primary consumer was intended to be `ztest` which only runs in user space so it was decided to not make this a valid option for the kernel modules.  A second example of this is `fletcher_4_impl_get()`.

**Comment by Teknoman117** [comment]
@behlendorf I believe the code coverage may be failing as the source code for the generic implementation is inactive on x86_64 for now due to an incompatibility with x86-64 (at least on AMD CPUs).  @tcaputi was able to reproduce what I was seeing.

**Comment by Teknoman117** [comment]
@tcaputi Still mulling over how best to deal with the fact that the generic
C implementation needs byte swapping but the accelerated implementations
don't need it.  Anyone care if I move the byte swapping logic (or lack
thereof) into the individual AES implementations?

**Comment by Teknoman117** [comment]
@tcaputi

I added a boolean to the ops struct which is set by ifdef's (per your suggestion).  If you set any mode (generic, x86_64, AES-NI) from the start (before loading any keys or mounting any encrypted datasets) and leave it that way for your session, all implementations can work with data generated by the other methods.

However, if you swap on the fly between one of the accelerated implementations (x86_64, AES-NI) to/from the generic implementation you get input/output errors (and sometimes kernel oopses when you try to export the pool after those errors).  Swapping between the accelerated implementations is fine (this also means the cycle setting causes problems).  Swapping between C implementation and the PCLMULQDQ implementations for the GCM method works fine and cycle causes no problems (other than costing you a lot of performance).

Are you aware of any differences in the key schedule that would cause these problems?

**Comment by tcaputi** [comment]
@Teknoman117 I'm not aware of any differences, but I'm sure there are some (since the 3 implementations I believe were written by 3 different people). This is what I was talking about in my earlier comment when I mentioned that an individual encryption / decryption session should not switch implementations during execution.

File: module/icp/include/aes/aes_impl.h [hunk 8]
```diff
@@ -106,17 +106,15 @@ typedef union {
...
-#define	INTEL_AES_NI_CAPABLE	0x1	/* AES-NI instructions present */
-
+typedef struct aes_impl_ops aes_impl_ops_t;
 typedef struct aes_key aes_key_t;
 struct aes_key {
 	aes_ks_t	encr_ks;  /* encryption key schedule */
 	aes_ks_t	decr_ks;  /* decryption key schedule */
 #ifdef __amd64
 	long double	align128; /* Align fields above for Intel AES-NI */
-	int		flags;	  /* implementation-dependent flags */
```
Comment (Teknoman117) [inline module/icp/include/aes/aes_impl.h]: This was added _specifically_ to hold the aes-ni support flag, which is no longer needed.

**Comment by Teknoman117** [comment]
@tcaputi 
I'm now keeping track of the aes backend in the aes key schedule structure, the thought for this being to use the implementation that generated the key for encryption and decryption with it.

@behlendorf 
I think I've addressed all of the existing concerns, if you could take another look at this that would be great.

File: module/icp/algs/aes/aes_impl.c [hunk 9]
```diff
@@ -1567,52 +211,214 @@ aes_alloc_keysched(size_t *size, int kmflag)
...
+aes_impl_set(const char *val)
+{
+	int err = -EINVAL;
+	char req_name[AES_IMPL_NAME_MAX];
+	uint32_t impl = AES_IMPL_READ(user_sel_impl);
+	size_t i;
+
+	/* sanitize input */
+	i = strnlen(val, AES_IMPL_NAME_MAX);
+	if (i == 0 || i == AES_IMPL_NAME_MAX)
```
Comment (tcaputi) [inline module/icp/algs/aes/aes_impl.c]: i >= AES_IMPL_NAME_MAX

File: module/icp/algs/aes/aes_impl.c [hunk 10]
```diff
@@ -1567,52 +211,214 @@ aes_alloc_keysched(size_t *size, int kmflag)
...
+		return (err);
+
+	strlcpy(req_name, val, AES_IMPL_NAME_MAX);
+	while (i > 0 && !!isspace(req_name[i-1]))
```
Comment (tcaputi) [inline module/icp/algs/aes/aes_impl.c]: nit: shouldn't need the `!!` here

File: module/icp/algs/modes/gcm.c [hunk 11]
```diff
@@ -694,55 +629,206 @@ gcm_set_kmflag(gcm_ctx_t *ctx, int kmflag)
...
-		} else {
-			cached_result = 0;
+	int err = -EINVAL;
+	char req_name[GCM_IMPL_NAME_MAX];
+	uint32_t impl = GCM_IMPL_READ(user_sel_impl);
+	size_t i;
+
+	/* sanitize input */
+	i = strnlen(val, GCM_IMPL_NAME_MAX);
+	if (i == 0 || i == GCM_IMPL_NAME_MAX)
```
Comment (tcaputi) [inline module/icp/algs/modes/gcm.c]: >=

File: module/icp/include/aes/aes_impl.h [hunk 12]
```diff
@@ -163,6 +161,50 @@ typedef enum aes_mech_type {
...
+typedef void		(*aes_dec_f)(const uint32_t[], int, const uint32_t[4],
+	uint32_t[4]);
+typedef boolean_t	(*aes_will_work_f)(void);
+
+#define	AES_IMPL_NAME_MAX (16)
+
+struct aes_impl_ops {
+	aes_gen_f gen;
+	aes_enc_f enc;
+	aes_dec_f dec;
```
Comment (tcaputi) [inline module/icp/include/aes/aes_impl.h]: can we use the full names here? ("generate", "encrypt", "decrypt")

File: module/icp/algs/aes/aes_impl.c [hunk 13]
```diff
@@ -1502,43 +153,36 @@ int
 aes_decrypt_block(const void *ks, const uint8_t *ct, uint8_t *pt)
 {
 	aes_key_t	*ksch = (aes_key_t *)ks;
+	const aes_impl_ops_t	*ops = aes_impl_get_ops();
```
Comment (tcaputi) [inline module/icp/algs/aes/aes_impl.c]: why is it that here we grab a new ops, when in `aes_encrypt()` we grab the one in the ksch? I'm worried about this because each encryption / decryption operation should use the same implementation from beginning to end.

File: include/linux/simd_x86.h [hunk 14]
```diff
@@ -442,6 +450,35 @@ zfs_bmi2_available(void)
 #endif
 }
 
+/*
+ * Check if AES instruction set is available
+ */
+static inline boolean_t
+zfs_aes_available(void)
```
Comment (tcaputi) [inline include/linux/simd_x86.h]: nit: can we rename this to `zfs_aes_instruction_available()` or something? This name makes it sound like encryption is completely impossible on this system.

File: module/icp/algs/aes/aes_impl_aesni.c [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+{
+	kfpu_begin();
+	key->nr = rijndael_key_setup_enc_intel(&(key->encr_ks.ks32[0]),
+	    keyarr32, keybits);
+	key->nr = rijndael_key_setup_dec_intel(&(key->decr_ks.ks32[0]),
+	    keyarr32, keybits);
+	kfpu_end();
+
+	key->type = AES_32BIT_KS;
+	key->ops = &aes_aesni_impl;
```
Comment (tcaputi) [inline module/icp/algs/aes/aes_impl_aesni.c]: I think this should be moved to `aes_keysched_init()`. Right now you grab a struct via the generic "get" function, which has a pointer to a function which sets a pointer back to the struct. Its a bit less confusing if all of the code that figures out what implementation is going to be used is all in one place.

File: include/linux/simd_x86.h (hunk 14 above)
Comment (behlendorf) [reply include/linux/simd_x86.h]: For consistentcy with the other instructions checks I'd prefer we leave the names as is.

File: module/icp/algs/modes/gcm.c (hunk 11 above)
Comment (Teknoman117) [reply module/icp/algs/modes/gcm.c]: did you mean >= ?

File: module/icp/algs/aes/aes_impl.c (hunk 13 above)
Comment (Teknoman117) [reply module/icp/algs/aes/aes_impl.c]: this is a mistake on my part, thanks for catching that.

File: module/icp/algs/modes/gcm.c (hunk 11 above)
Comment (tcaputi) [reply module/icp/algs/modes/gcm.c]: yes. Sorry. it put a space there and so it became a quote.

File: module/icp/algs/aes/aes_impl.c [hunk 16]
```diff
@@ -1567,52 +211,214 @@ aes_alloc_keysched(size_t *size, int kmflag)
...
+		static size_t cycle_impl_idx = 0;
+		size_t idx = (++cycle_impl_idx) % aes_supp_impl_cnt;
+		ops = aes_supp_impl[idx];
+	}
+	break;
+	default:
+		ASSERT3U(impl, <, aes_supp_impl_cnt);
+		ASSERT3U(aes_supp_impl_cnt, >, 0);
+		if (impl < ARRAY_SIZE(aes_all_impl))
+			ops = aes_supp_impl[impl];
```
[the remark this answers is not in the document]
Comment (Teknoman117) [reply module/icp/algs/aes/aes_impl.c]: This assert is there because ARRAY_SIZE(aes_all_impl) is probably greater than aes_supp_impl_cnt.  Hypothetically if some strange error occurred while setting the implementation, impl could be greater than aes_supp_impl_cnt but less than ARRAY_SIZE(aes_all_impl), causing you to get a bogus pointer (in c99 static fixed size arrays (i.e. aes_all_impl) are zeroed, but I have no idea for the gnu89 spec the kernel uses).

If we wanted to remove an assertion, I could change the if statement to use aes_supp_impl_cnt instead of ARRAY_SIZE(aes_all_impl), as aes_supp_impl_cnt will always be less than ARRAY_SIZE(aes_all_impl) save for some cosmic ray bit flipping.

File: module/icp/algs/aes/aes_impl_aesni.c (hunk 15 above)
Comment (Teknoman117) [reply module/icp/algs/aes/aes_impl_aesni.c]: I understand your thoughts on moving where key->ops get set, but while there aren't any code paths currently that use the 64 bit key schedule, couldn't there theoretically be some in the future?  Shouldn't we leave the type setting in the backend implementations?

Comment (tcaputi) [reply module/icp/algs/aes/aes_impl_aesni.c]: I generally follow the rule of 3s on these things. We have 4 implementations of AES  in ZoL (including QAT) and none of them use it. I say its safe to make this change.

File: module/icp/algs/aes/aes_impl.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (tcaputi) [reply module/icp/algs/aes/aes_impl.c]: ok. in that case its probably fine as it is.

**Comment by Teknoman117** [comment]
Sorry for the long delay again.  I've addressed ztest and @tcaputi nits.

File: module/icp/algs/aes/aes_impl.c [hunk 17]
```diff
@@ -1567,52 +219,214 @@ aes_alloc_keysched(size_t *size, int kmflag)
...
+	/* move supported impl into aes_supp_impls */
+	for (i = 0, c = 0; i < ARRAY_SIZE(aes_all_impl); i++) {
+		curr_impl = (aes_impl_ops_t *)aes_all_impl[i];
+
+		if (curr_impl->is_supported())
+			aes_supp_impl[c++] = (aes_impl_ops_t *)curr_impl;
+	}
+	aes_supp_impl_cnt = c;
+
+	/* set fastest implementation. assume hardware accelerated is fastest */
```
Comment (behlendorf) [inline module/icp/algs/aes/aes_impl.c]: Based on the local testing I've done with this change it seems we can't assume the hardware accelerated version will always be fastest.  Using a dual-socket "AMD EPYC 7451 24-Core Processor" I was getting the best performance with `icp_aes_impl=x86_64` followed by the `generic` and then the `aesni` version was bringing up the rear at a little better than _half_ the x86_64 rate.

@Teknoman117 can you look in to adding a micro-benchmark similar to the RAIDZ code to dynamically determine the fastest implementation during initiailization.

**Comment by Teknoman117** [comment]
I noticed a sharp drop in the performance gain I got sometime early March,
so I need to bisect to see if something happened (I also switched my
threadripper from interleaved (virtual UMA basically) into NUMA mode
sometime in there so it might be that as well).  I do have access to some
EPYC systems for testing.

Can you check what gains you get with switching away from fastest in gcm
mode?  I saw the most improvement there (pclmulqdq being 10x the
performance of generic on my threadripper).

**Comment by behlendorf** [comment]
@Teknoman117 were you able to retest this and determine why you're seeing less of a performance win.  This slipped off my list but time permitting I'll re-check the performance I'm getting.

**Comment by Teknoman117** [comment]
@behlendorf I wasn't able to confirm the accelerated version being slower on my own box.  Got access to a couple of EPYC boxes so I'll be trying to test some of these.

**Comment by behlendorf** [comment]
@Teknoman117 when you get a moment can you please rebase this on master.  I'll see about reproducing the performance numbers I observed.

**Comment by behlendorf** [comment]
@Teknoman117 thanks.  I did a little benchmarking and analysis of the updated PR using the basic `dd` read test from your original comment. It was easiest to do a survey of all the `icp_aes_impl` and `icp_aes_impl` settings so that's what I ran.  I'll see about doing some testing with `fio` too.

System:
- AMD EPYC 7451 24-Core Processor
- 2.3GHz base and 2.9GHz turbo
- 128G memory
- 64G ramdisk for testing.

Configuration:

```
NAME   SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
test  63.5G  32.0G  31.5G        -         -     0%    50%  1.00x  ONLINE  -
  ram0  63.5G  32.0G  31.5G      -         -     0%    50%
```

Test:
- Using a single `dd` read a cold uncached 32G encrypted file from disk
  `dd if=/mnt/encrypted/file.bin of=/dev/null bs=1024k count=32768`
- Pre-cached read speed: __4.0 GB/s__
- Tested `encryption=aes-256-ccm` and `encryption=aes-256-gcm`
- All results in MiB/s

encryption=aes-256-ccm | generic | x86_64 | aesni
--- | --- | --- | ---
generic | 513 | 654 | 403
pclmulqdq |  523 | 654 | 407

encryption=aes-256-gcm | generic | x86_64 | aesni
--- | --- | --- | ---
generic | 233 | 242 | 219
pclmulqdq |  492 | 545 | 431

On the surface my performance numbers aren't as impressive as your results, but this is almost certainly due to the EPYC's lower cpu frequency than your Threadripper.  There is still a nice performance win.  But more importantly `perf` clearly shows cpu is greatly reduced as expected.  The sustained usage for the critical functions was:

function | percent cpu
--- | ---
gcm_generic_mul | 55.3%
gcm_mul_pclmulqdq | 1.7%
aes_generic_encrypt | 17.8%
aes_encrypt_amd64 | 12.3%
aes_encrypt_intel | 2.1%

Bottom line, this looks good.  I'll put it though some additional torture tests and throw some parallel workloads at it but I don't expect to see any issues.  Give my initial results we could probably hold off on the micro benchmarks if you can't get to them.  We can always follow up this PR with another to add them if we determine they are need for some hardware.

**Comment by Teknoman117** [comment]
When do you guys think a release is going to be cut that includes encryption?

**Comment by Skaronator** [comment]
0.8 will include encryption. There is no ETA for a release (it's done when it's done) but my guess is RC for 2019 and release in 2020 (or end of 2019).

**Comment by behlendorf** [comment]
@Teknoman117 expect the first release candidate to be tagged after allocation classes is merged.  Which is expected to be wrapped up in the next month or so.
